File: apps/web-admin/src/app/features/map/utils/map-motes-effect.ts
import type { AmbientTickOptions } from './map-ambient-zone';
import { parkParticleTarget } from './map-park-visual-scale';
import { ambientScreenScale } from './map-ambient-zone';
import { ambientWindVector, DEFAULT_AMBIENT_WIND } from './map-ambient-wind';

interface Mote {
  bx: number;
  by: number;
  vx: number;
  vy: number;
  r: number;
  phase: number;
  warm: boolean;
}

/** Partículas flotantes (polen, polvo en luz, chispas suaves). */
export class MapMotesEffect {
  private motes: Mote[] = [];
  private intensity = 0.4;
  private sizeMul = 1;
  private containsPoint: ((bx: number, by: number) => boolean) | null = null;

  setIntensity(value: number): void {
    this.intensity = Math.min(1, Math.max(0, value));
  }

  setSizeMul(value: number): void {
    this.sizeMul = Math.min(2.5, Math.max(0.08, value));
  }

  setContainsPoint(fn: ((bx: number, by: number) => boolean) | null): void {
    this.containsPoint = fn;
  }

  clear(): void {
    this.motes = [];
  }

  private inZone(bx: number, by: number): boolean {
    return !this.containsPoint || this.containsPoint(bx, by);
  }

  private spawn(bounds: AmbientTickOptions['bounds'], wind = DEFAULT_AMBIENT_WIND): Mote {
    const spanX = bounds.maxX - bounds.minX;
    const spanY = bounds.maxY - bounds.minY;
    let bx = bounds.minX + Math.random() * spanX;
    let by = bounds.minY + Math.random() * spanY;
    for (let t = 0; t < 20; t++) {
      bx = bounds.minX + Math.random() * spanX;
      by = bounds.minY + Math.random() * spanY;
      if (this.inZone(bx, by)) break;
    }
    const wv = ambientWindVector(wind);
    return {
      bx,
      by,
      vx: wv.vx * 0.22 + (Math.random() - 0.5) * 0.18,
      vy: wv.vy * 0.18 - 0.12 - Math.random() * 0.25,
      r: 0.28 + Math.random() * 0.65,
      phase: Math.random() * Math.PI * 2,
      warm: Math.random() > 0.55,
    };
  }

  tick(options: AmbientTickOptions, dt = 1): void {
    if (options.containsPoint) this.containsPoint = options.containsPoint;
    const wind = options.wind ?? DEFAULT_AMBIENT_WIND;
    const bounds = options.bounds;
    const spanX = bounds.maxX - bounds.minX;
    const spanY = bounds.maxY - bounds.minY;
    if (spanX <= 0 || spanY <= 0) return;

    const inten = Math.max(0.15, this.intensity);
    const target = parkParticleTarget(8, 28, inten);

    while (this.motes.length < target) this.motes.push(this.spawn(bounds, wind));
    while (this.motes.length > target) this.motes.pop();

    for (const m of this.motes) {
      m.phase += 0.04 * dt;
      m.bx += m.vx * dt + Math.sin(m.phase) * 0.06 * dt;
      m.by += m.vy * dt;

      if (!this.inZone(m.bx, m.by) || m.by < bounds.minY - 30 || m.by > bounds.maxY + 20) {
        const n = this.spawn(bounds, wind);
        Object.assign(m, n);
        m.by = bounds.maxY + Math.random() * spanY * 0.15;
      }
    }
  }

  draw(
    ctx: CanvasRenderingContext2D,
    clipPath: Path2D | null,
    toScreen: (bx: number, by: number) => { x: number; y: number },
    screenScale = 1,
  ): void {
    const inten = Math.max(0.2, this.intensity);
    const sr = ambientScreenScale(screenScale, this.sizeMul);
    ctx.save();
    if (clipPath) ctx.clip(clipPath);

    for (const m of this.motes) {
      if (!this.inZone(m.bx, m.by)) continue;
      const tw = 0.45 + Math.sin(m.phase) * 0.55;
      const alpha = inten * tw;
      const { x, y } = toScreen(m.bx, m.by);
      const r = Math.max(0.25, m.r * sr * 0.9);

      if (m.warm) {
        ctx.fillStyle = `rgba(255, 230, 160, ${alpha * 0.9})`;
      } else {
        ctx.fillStyle = `rgba(210, 245, 255, ${alpha * 0.75})`;
      }
      ctx.beginPath();
      ctx.arc(x, y, r, 0, Math.PI * 2);
      ctx.fill();

      ctx.fillStyle = `rgba(255, 255, 255, ${alpha * 0.35})`;
      ctx.beginPath();
      ctx.arc(x - r * 0.25, y - r * 0.25, r * 0.35, 0, Math.PI * 2);
      ctx.fill();
    }

    ctx.restore();
  }
}

File: apps/mobile-android/app/src/main/java/com/univalle/pedrochacolla/ui/dashboard/MapMotesEffect.kt
package com.univalle.pedrochacolla.ui.dashboard

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RadialGradient
import android.graphics.Shader
import com.univalle.pedrochacolla.utils.map.AmbientTickOptions
import com.univalle.pedrochacolla.utils.map.AmbientWind
import com.univalle.pedrochacolla.utils.map.MapAmbientZone
import kotlin.math.sin
import kotlin.random.Random

class MapMotesEffect {
    var intensity = 0.4f
    var sizeMul = 1f
        set(value) { field = value.coerceIn(0.08f, 2.5f) }

    private data class Mote(
        var bx: Float, var by: Float, var vx: Float, var vy: Float,
        val r: Float, var phase: Float, val warm: Boolean,
    )

    private val motes = mutableListOf<Mote>()
    private var containsPoint: ((Float, Float) -> Boolean)? = null
    private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE }

    fun clear() { motes.clear() }
    fun setContainsPoint(fn: ((Float, Float) -> Boolean)?) { containsPoint = fn }

    private fun inZone(bx: Float, by: Float) = containsPoint?.invoke(bx, by) ?: true

    private fun spawn(bounds: AmbientTickOptions, wind: AmbientWind): Mote {
        val b = bounds.bounds
        val spanX = b.maxX - b.minX
        val spanY = b.maxY - b.minY
        var bx = b.minX + Random.nextFloat() * spanX
        var by = b.minY + Random.nextFloat() * spanY
        repeat(20) {
            bx = b.minX + Random.nextFloat() * spanX
            by = b.minY + Random.nextFloat() * spanY
            if (inZone(bx, by)) return@repeat
        }
        val (wx, wy) = MapAmbientZone.ambientWindVector(wind)
        return Mote(
            bx, by,
            wx * 0.22f + (Random.nextFloat() - 0.5f) * 0.18f,
            wy * 0.18f - 0.12f - Random.nextFloat() * 0.25f,
            0.28f + Random.nextFloat() * 0.65f,
            Random.nextFloat() * 6.28f,
            Random.nextFloat() > 0.55f,
        )
    }

    fun tick(options: AmbientTickOptions, dt: Float = 1f) {
        options.containsPoint?.let { containsPoint = it }
        val b = options.bounds
        val spanX = b.maxX - b.minX
        val spanY = b.maxY - b.minY
        if (spanX <= 0f || spanY <= 0f) return
        val inten = maxOf(0.15f, intensity)
        val target = MapAmbientZone.parkParticleTarget(8, 28, inten)
        while (motes.size < target) motes.add(spawn(options, options.wind))
        while (motes.size > target) motes.removeAt(motes.lastIndex)
        for (m in motes) {
            m.phase += 0.04f * dt
            m.bx += m.vx * dt + sin(m.phase) * 0.06f * dt
            m.by += m.vy * dt
            if (!inZone(m.bx, m.by) || m.by < b.minY - 30f || m.by > b.maxY + 20f) {
                val n = spawn(options, options.wind)
                m.bx = n.bx; m.by = b.maxY + Random.nextFloat() * spanY * 0.15f
                m.vx = n.vx; m.vy = n.vy; m.phase = n.phase
            }
        }
    }

    fun draw(canvas: Canvas, clipPath: Path?, toScreen: (Float, Float) -> Pair<Float, Float>, screenScale: Float) {
        val inten = maxOf(0.2f, intensity)
        val sr = MapAmbientZone.ambientScreenScale(screenScale, sizeMul)
        canvas.save()
        clipPath?.let { canvas.clipPath(it) }
        for (m in motes) {
            if (!inZone(m.bx, m.by)) continue
            val tw = 0.45f + sin(m.phase) * 0.55f
            val alpha = (inten * tw).coerceIn(0f, 1f)
            val (x, y) = toScreen(m.bx, m.by)
            val r = maxOf(0.25f, m.r * sr * 0.9f)
            val haloR = r * if (m.warm) 4.5f else 3f
            val haloA = (alpha * if (m.warm) 0.5f else 0.3f * 255).toInt()
            fillPaint.shader = RadialGradient(
                x, y, haloR,
                intArrayOf(
                    if (m.warm) android.graphics.Color.argb(haloA, 255, 224, 150)
                    else android.graphics.Color.argb(haloA, 200, 240, 255),
                    android.graphics.Color.TRANSPARENT,
                ),
                floatArrayOf(0f, 1f),
                Shader.TileMode.CLAMP,
            )
            canvas.drawCircle(x, y, haloR, fillPaint)
            fillPaint.shader = null
            val bodyR = r * 1.15f
            fillPaint.color = if (m.warm) android.graphics.Color.argb((alpha * 255).toInt(), 255, 206, 70)
            else android.graphics.Color.argb((alpha * 255).toInt(), 150, 220, 255)
            canvas.drawCircle(x, y, bodyR, fillPaint)
            strokePaint.color = if (m.warm) android.graphics.Color.argb((alpha * 255).toInt(), 120, 70, 0)
            else android.graphics.Color.argb((alpha * 255).toInt(), 20, 60, 110)
            strokePaint.strokeWidth = maxOf(0.3f, bodyR * 0.28f)
            canvas.drawCircle(x, y, bodyR, strokePaint)
            fillPaint.color = android.graphics.Color.argb((alpha * 255).toInt(), 255, 255, 255)
            canvas.drawCircle(x - bodyR * 0.15f, y - bodyR * 0.15f, bodyR * 0.45f, fillPaint)
        }
        canvas.restore()
    }
}
